Refactor to use ScrapedPage subclasses

chrismytton opened the pull request:
This changes the scraper to use `ScrapedPage` subclasses, which now handle stripping the session information from the url.

# Notes to reviewer

You can see the scraper doing things by running the following:

    VERBOSE=1 bundle exec ruby scraper.rb

This is currently missing the archiver because I've hit a couple of `scraped_page_archive` bugs relating to VCR (I think because the archive branch on this repo contains some non-standard cassettes, so it was a good test-case!).

# Notes to merger

Set https://morph.io/everypolitician-scrapers/spain_congreso_es to auto-run once this is merged.

chrismytton committed:
Add scraped_page dependency

chrismytton committed:
WIP: Refactor scraper to use ScrapedPage classes

chrismytton committed:
Add class to represent members' DOB

chrismytton committed:
Add party field

chrismytton committed:
Add remaining methods to MemberPage class

These pick up the other fields that the existing scraper was getting.

chrismytton committed:
Move DateOfBirth class into its own file

chrismytton committed:
Handle missing faction information

chrismytton committed:
More robust url session handling

Use the URI module to make it easier to just replace the unwanted part
of the query string.

chrismytton committed:
Extract a SpanishCongressPage base class

This means that rather than having to override the url in every class in
the system we can just do it once here and then inherit from this class.

chrismytton committed:
Move String#tidy into its own file

This should become part of ScapedPage RSN.

chrismytton committed:
Make MembersListPage inherit from SpanishCongressPage

This means these pages also have the session information stripped from
the url.

chrismytton committed:
Handle pagination in scraper.rb

Rather than just scraping the first page, scrape all pages.

chrismytton committed:
Convert name parts to string before tidying

chrismytton committed:
Correctly handle the last page of results

On the last page there is no next link, so trying to call `next[:href]`
blows up. Instead I've moved `next_page_link` to a separate method then
checking that exists in `next_page_url`.

chrismytton committed:
Simplify phone and fax fields

chrismytton committed:
DateOfBirth#to_s should always return a string

chrismytton committed:
Run rubocop -a on lib

chrismytton committed:
Delete unused code in scraper.rb

chrismytton committed:
Remove unneeded gems

Most gems we need now come as dependencies of scraped_page.

chrismytton commented:
@tmtmtmtm 👀

tmtmtmtm commented on Gemfile:
```diff
@@ -8,10 +8,5 @@ git_source(:github) { |repo_name| "https://github.com/#{repo_name}.git" }
-gem "capybara"
-gem "poltergeist"
-gem 'scraped_page_archive', github: "everypolitician/scraped_page_archive", branch: "master"
+gem 'scraped_page', github: 'everypolitician/scraped_page', branch: 'master'
```
Perhaps worth running a rubocop tidy against this file too, so we don't have mismatched quotes like this?

tmtmtmtm commented on lib/member_page.rb:
```diff
@@ -0,0 +1,115 @@
+
+class MemberPage < SpanishCongressPage
+  field :iddiputado do
+    query['idDiputado']
```
`query` is a little more ambiguous in this context than when you're already explicitly dealing with a URL. Perhaps something like `query_string` would be a little more obvious?

chrismytton replied:
Yeah good point, I think `query_string` is much clearer, will change it to that.

tmtmtmtm commented on lib/member_page.rb:
```diff
@@ -0,0 +1,115 @@
+  end
+
+  def group
+    @group ||= noko.at_css('div#curriculum div.texto_dip ul li div.dip_rojo:last').text.tidy
```
is it really worth memoizing all these things? I don't think they're particularly costly, and some of them aren't even called more than once anyway.

chrismytton replied:
No, just habit, but you're quite right, this is a form of unnecessary premature optimization.

tmtmtmtm commented on lib/members_list_page.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+require_relative 'spanish_congress_page'
+
+class MembersListPage < SpanishCongressPage
```
A 'ScrapedPage' with no fields is a bit of a red flag. The first two of these look like they should be fields to me (and the `next_page_link` a private method).

tmtmtmtm reviewed (changes requested):
Some comments in place, but the bigger problem here is that this doesn't seem to be fetching the history for each person.

Each person page has a 'Todas las legislaturas' link that goes to a page that lists all the different term memberships they've had: e.g. http://www.congreso.es/portal/page/portal/Congreso/Congreso/Diputados/BusqForm?idDiputado=174&idLegislatura=7

chrismytton committed:
Add open-uri-cached

This caches a copy of fetched webpages to make the scraper quicker to
develop locally.
